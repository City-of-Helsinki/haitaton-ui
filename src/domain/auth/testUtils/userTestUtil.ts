import { Profile, User } from 'hds-react';

const authority = 'https://api.hel.fi/sso/openid';
const client_id = 'test-client';

const tokenExpirationTimeInSeconds = 3600;

export function createUser(placeUserToStorage = true, userProfile?: Partial<Profile>): User {
  const nowAsSeconds = Math.round(Date.now() / 1000);
  const expires_in = tokenExpirationTimeInSeconds;
  const expires_at = nowAsSeconds + expires_in;

  const user: User = {
    access_token: placeUserToStorage ? 'access_token' : '',
    expires_at,
    id_token: 'id_token',
    profile: {
      sub: 'sub',
      iss: 'issuer',
      aud: 'aud',
      exp: expires_at,
      iat: nowAsSeconds,
      given_name: 'Test',
      family_name: 'User',
      name: 'Test User',
      email: 'test.user@mail.com',
      amr: ['validAmr'],
      ...userProfile,
    },
    refresh_token: 'refresh_token',
    scope: 'openid profile',
    session_state: String(`${Math.random()}${Math.random()}`),
    state: '',
    token_type: 'Bearer',
    expires_in,
    expired: false,
    scopes: ['openid', 'profile'],
    toStorageString() {
      return JSON.stringify(this);
    },
  };

  if (placeUserToStorage) {
    sessionStorage.setItem(`oidc.user:${authority}:${client_id}`, user.toStorageString());
  }

  return user;
}
