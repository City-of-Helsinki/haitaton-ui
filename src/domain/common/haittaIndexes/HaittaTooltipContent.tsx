import { Trans, useTranslation } from 'react-i18next';
import { Box } from '@chakra-ui/react';

type Props = {
  translationKey: string;
  showHeading?: boolean;
};

export default function HaittaTooltipContent({
  translationKey,
  showHeading = true,
}: Readonly<Props>) {
  const { t } = useTranslation();

  return (
    <>
      {showHeading && (
        <Box as="h5" fontWeight="bold">
          {t('hankeIndexes:haittaindeksit')}:
        </Box>
      )}
      <Trans
        i18nKey={translationKey}
        components={{ ul: <Box as="ul" listStyleType="none" />, li: <li /> }}
      />
    </>
  );
}
