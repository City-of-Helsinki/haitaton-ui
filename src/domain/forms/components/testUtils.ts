import { fireEvent, screen } from '@testing-library/dom';

export function fillNewContactPersonForm(
  options: {
    etunimi?: string;
    sukunimi?: string;
    sahkoposti?: string;
    puhelinnumero?: string;
  } = {},
) {
  const {
    etunimi = 'Matti',
    sukunimi = 'Meikäläinen',
    sahkoposti = 'matti.meikalainen@test.com',
    puhelinnumero = '0401234567',
  } = options;
  fireEvent.change(screen.getByLabelText(/etunimi/i), {
    target: { value: etunimi },
  });
  fireEvent.change(screen.getByLabelText(/sukunimi/i), {
    target: { value: sukunimi },
  });
  fireEvent.change(screen.getAllByRole('textbox', { name: /sähköposti/i })[1], {
    target: { value: sahkoposti },
  });
  fireEvent.change(screen.getAllByRole('textbox', { name: /puhelin/i })[1], {
    target: { value: puhelinnumero },
  });
}
