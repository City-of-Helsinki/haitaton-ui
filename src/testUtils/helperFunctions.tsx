import { fireEvent, RenderResult, screen, waitForElementToBeRemoved } from '@testing-library/react';

export const changeFilterDate = (
  label: string,
  renderedComponent: RenderResult,
  value: string | null,
) => {
  fireEvent.change(renderedComponent.getByLabelText(label, { exact: false }), {
    target: { value },
  });
};

export function waitForLoadingToFinish() {
  return waitForElementToBeRemoved(() => screen.queryByText(/page is loading/i), {
    timeout: 4000,
  });
}
