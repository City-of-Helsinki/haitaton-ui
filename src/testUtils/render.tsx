import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import { QueryClient, QueryClientProvider } from 'react-query';
import { render, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { I18nextProvider } from 'react-i18next';
import { BrowserRouter } from 'react-router-dom';
import i18n from '../locales/i18nForTests';
import { store } from '../common/redux/store';
import { GlobalNotificationProvider } from '../common/components/globalNotification/GlobalNotificationContext';
import GlobalNotification from '../common/components/globalNotification/GlobalNotification';

type Props = {
  children: React.ReactChildren;
};

const AllTheProviders = ({ children }: Props) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retryDelay: 0,
      },
    },
  });

  return (
    <BrowserRouter>
      <ReduxProvider store={store}>
        <QueryClientProvider client={queryClient}>
          <I18nextProvider i18n={i18n}>
            <GlobalNotificationProvider>
              {children}
              <GlobalNotification />
            </GlobalNotificationProvider>
          </I18nextProvider>
        </QueryClientProvider>
      </ReduxProvider>
    </BrowserRouter>
  );
};

const customRender = (ui: React.ReactElement, options: RenderOptions = {}, route = '/') => {
  window.history.pushState({}, 'Test page', route);
  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: AllTheProviders as React.ComponentType, ...options }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
