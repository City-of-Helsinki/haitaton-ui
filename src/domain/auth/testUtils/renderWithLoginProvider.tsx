import { render } from '@testing-library/react';
import {
  Beacon,
  ConnectedModule,
  LoginProvider,
  OidcClient,
  OidcClientError,
  OidcClientProps,
  OidcClientState,
  triggerForAllOidcClientSignals,
} from 'hds-react';
import { createUser } from './userTestUtil';
import userEvent from '@testing-library/user-event';

const authority = 'https://api.hel.fi/sso/openid';
const client_id = 'test-client';
const scope = 'openid profile';
const oidcClientTestProps: OidcClientProps = {
  userManagerSettings: {
    authority,
    client_id,
    scope,
  },
};

export type RenderWithLoginProviderProps = {
  state: OidcClientState;
  returnUser: boolean;
  placeUserToStorage?: boolean;
  userADGroups?: string[];
  errorType?: 'SIGNIN_ERROR' | 'INVALID_OR_EXPIRED_USER' | 'RENEWAL_FAILED';
  children?: React.ReactNode;
};

export function renderWithLoginProvider({
  state,
  returnUser,
  placeUserToStorage = true,
  userADGroups,
  errorType,
  children,
}: RenderWithLoginProviderProps) {
  let beacon: Beacon;
  const handleError = new OidcClientError('Handlecallback failed', errorType ?? 'SIGNIN_ERROR');
  const helperModule: ConnectedModule = {
    namespace: 'helper',
    connect: (targetBeacon) => {
      beacon = targetBeacon;
      beacon.addListener(triggerForAllOidcClientSignals, (signal) => {
        const user = createUser(placeUserToStorage, userADGroups);
        const oidcClient = signal.context as OidcClient;
        jest.spyOn(oidcClient, 'getState').mockReturnValue(state);
        jest.spyOn(oidcClient, 'getUser').mockReturnValue(user);
        if (!returnUser) {
          jest.spyOn(oidcClient, 'handleCallback').mockRejectedValue(handleError);
        } else {
          jest.spyOn(oidcClient, 'handleCallback').mockResolvedValue(user);
        }
      });
    },
  };

  return {
    user: userEvent.setup(),
    ...render(
      <LoginProvider {...oidcClientTestProps} modules={[helperModule]}>
        {children}
      </LoginProvider>,
    ),
  };
}
