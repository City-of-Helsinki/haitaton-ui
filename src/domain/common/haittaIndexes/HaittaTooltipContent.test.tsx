import { render, screen } from '../../../testUtils/render';
import HaittaTooltipContent from './HaittaTooltipContent';

test.each([
  ['hankeIndexes:tooltips:PYORALIIKENNE'],
  ['hankeIndexes:tooltips:AUTOLIIKENNE'],
  ['hankeIndexes:tooltips:LINJAAUTOLIIKENNE'],
  ['hankeIndexes:tooltips:RAITIOLIIKENNE'],
  ['hankeIndexes:tooltips:MUUT'],
])('Should show correct tooltip for %s', (translationKey) => {
  const { container } = render(<HaittaTooltipContent translationKey={translationKey} />);
  expect(container).toMatchSnapshot();
});

test('Should not show heading when showHeading is false', () => {
  render(
    <HaittaTooltipContent
      translationKey="hankeIndexes:tooltips:PYORALIIKENNE"
      showHeading={false}
    />,
  );
  expect(screen.queryByText('Haittaindeksit:')).not.toBeInTheDocument();
});
