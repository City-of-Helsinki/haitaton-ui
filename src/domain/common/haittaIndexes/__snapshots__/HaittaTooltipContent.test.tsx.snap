// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should show correct tooltip for hankeIndexes:tooltips:AUTOLIIKENNE 1`] = `
<div>
  <h5
    class="css-in3yi3"
  >
    Haittaindeksit
    :
  </h5>
  Autoliikenteelle lasketaan painotettu haittaindeksi (0-5) siihen sisältyvien osa-alueiden mukaan.
</div>
`;

exports[`Should show correct tooltip for hankeIndexes:tooltips:LINJAAUTOLIIKENNE 1`] = `
<div>
  <h5
    class="css-in3yi3"
  >
    Haittaindeksit
    :
  </h5>
  <ul
    class="css-155za0w"
  >
    <li>
      0: Ei linja-autojen paikallisliikennettä
    </li>
    <li>
      2: Linja-autojen paikallisliikennettä muulloin, kuin ruuhka-aikana
    </li>
    <li>
      3: Linja-autojen paikallisliikenteen linjojen vuoromäärä enintään 10 ruuhkatunnissa
    </li>
    <li>
      4: Linja-autojen paikallisliikenteen runkolinja tai vuoromäärä enintään 20 ruuhkatunnissa
    </li>
    <li>
      5: Kamppi-Rautatientori -alue, Mannerheimintie, Kaisaniemenkatu, Hämeentie tai linja-autojen paikallisliikenteen linjojen vuoromäärä yli 20 ruuhkatunnissa
    </li>
  </ul>
</div>
`;

exports[`Should show correct tooltip for hankeIndexes:tooltips:MUUT 1`] = `
<div>
  <h5
    class="css-in3yi3"
  >
    Haittaindeksit
    :
  </h5>
  key 'tooltips.MUUT (fi)' returned an object instead of string.
</div>
`;

exports[`Should show correct tooltip for hankeIndexes:tooltips:PYORALIIKENNE 1`] = `
<div>
  <h5
    class="css-in3yi3"
  >
    Haittaindeksit
    :
  </h5>
  <ul
    class="css-155za0w"
  >
    <li>
      0: Ei tunnistettua pyöräliikenteen reittiä
    </li>
    <li>
      2: Muut pyöräliikenteen yhteydet
    </li>
    <li>
      3: Muut pyöräliikenteen reitit (tai reitin/yhteyden luokitustieto puuttuu pyöräliikenteen tausta-aineistosta)
    </li>
    <li>
      5: Pyöräliikenteen pääreitti tai pääreitin osana toimiva katu
    </li>
  </ul>
</div>
`;

exports[`Should show correct tooltip for hankeIndexes:tooltips:RAITIOLIIKENNE 1`] = `
<div>
  <h5
    class="css-in3yi3"
  >
    Haittaindeksit
    :
  </h5>
  <ul
    class="css-155za0w"
  >
    <li>
      0: Ei tunnistettuja raitiotiekiskoja
    </li>
    <li>
      3: Raitiotieverkon rataosa, joilla ei ole säännöllistä linjaliikennettä
    </li>
    <li>
      5: Raitiotieverkon rataosa, jolla on säännöllistä linjaliikennettä
    </li>
  </ul>
</div>
`;
